### PR TITLE
SDKS-725 WebAuth Implementation

### PR DESCRIPTION
**Pull request by witrisna**

### CODE REVIEW

File: forgerock-auth-ui/src/main/java/org/forgerock/android/auth/ui/CallbackFragmentFactory.java [hunk 1]
```diff
@@ -59,6 +59,8 @@ private CallbackFragmentFactory() {
         register(ReCaptchaCallback.class, ReCaptchaCallbackFragment.class);
         register(ConsentMappingCallback.class, ConsentMappingCallbackFragment.class);
         register(DeviceProfileCallback.class, DeviceProfileCallbackFragment.class);
+        register(WebAuthnRegistrationCallback.class, WebAuthnRegistrationCallbackFragment.class);
```
Comment (rodrigoareis) [inline forgerock-auth-ui/src/main/java/org/forgerock/android/auth/ui/CallbackFragmentFactory.java:62]: Update copyright

File: forgerock-auth/src/androidTest/java/org/forgerock/android/auth/AuthInstrumentTestSuite.java [hunk 2]
```diff
@@ -9,6 +9,8 @@
 
 import org.forgerock.android.auth.callback.DeviceProfileCollectorCallbackAndroidTest;
 import org.forgerock.android.auth.collector.FRDeviceIdentifierTest;
+import org.forgerock.android.auth.webauthn.WebAuthnAuthenticationAndroidTest;
```
Comment (rodrigoareis) [inline forgerock-auth/src/androidTest/java/org/forgerock/android/auth/AuthInstrumentTestSuite.java:12]: Update copyright

File: forgerock-auth/src/androidTest/java/org/forgerock/android/auth/DefaultSingleSignOnManagerTest.java [hunk 3]
```diff
@@ -31,6 +31,7 @@
 @RunWith(AndroidVersionAwareTestRunner.class)
 public class DefaultSingleSignOnManagerTest {
 
+    public static final String ORG_FORGEROCK = "org.forgerock.test";
```
Comment (rodrigoareis) [inline forgerock-auth/src/androidTest/java/org/forgerock/android/auth/DefaultSingleSignOnManagerTest.java:34]: Update copyright

File: forgerock-auth/src/androidTest/java/org/forgerock/android/auth/MockServer.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+public abstract class MockServer {
+
+    protected MockWebServer server;
+
+    public Context context = ApplicationProvider.getApplicationContext();
+
+    @Before
+    public void startServer() throws Exception {
+
```
Comment (rodrigoareis) [inline forgerock-auth/src/androidTest/java/org/forgerock/android/auth/MockServer.java:32]: Suggest remove extra empty lines to be consistent with the coding style in the rest of the class.

File: forgerock-auth/src/androidTest/java/org/forgerock/android/auth/MockServer.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+import com.squareup.okhttp.mockwebserver.MockWebServer;
+
+import org.apache.commons.io.IOUtils;
+import org.junit.After;
+import org.junit.Before;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+
+public abstract class MockServer {
```
Comment (rodrigoareis) [inline forgerock-auth/src/androidTest/java/org/forgerock/android/auth/MockServer.java:24]: Missing javadoc

File: forgerock-auth-ui/src/main/java/org/forgerock/android/auth/ui/callback/WebAuthnAuthenticationCallbackFragment.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import org.forgerock.android.auth.callback.WebAuthnAuthenticationCallback;
+import org.forgerock.android.auth.ui.R;
+
+/**
+ * UI representation for {@link WebAuthnAuthenticationCallback}
+ */
+public class WebAuthnAuthenticationCallbackFragment extends CallbackFragment<WebAuthnAuthenticationCallback> {
+    private ProgressBar progressBar;
+
+    public WebAuthnAuthenticationCallbackFragment() {
```
Comment (rodrigoareis) [inline forgerock-auth-ui/src/main/java/org/forgerock/android/auth/ui/callback/WebAuthnAuthenticationCallbackFragment.java:26]: Should we provide a javadoc for this constructor?

File: forgerock-auth-ui/src/main/java/org/forgerock/android/auth/ui/callback/WebAuthnRegistrationCallbackFragment.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+import static android.view.View.GONE;
+
+/**
+ * UI representation for {@link WebAuthnRegistrationCallback}
+ */
+public class WebAuthnRegistrationCallbackFragment extends CallbackFragment<WebAuthnRegistrationCallback> {
+
+    private ProgressBar progressBar;
+
+    public WebAuthnRegistrationCallbackFragment() {
```
Comment (rodrigoareis) [inline forgerock-auth-ui/src/main/java/org/forgerock/android/auth/ui/callback/WebAuthnRegistrationCallbackFragment.java:29]: Should we provide a javadoc for this constructor?

File: forgerock-auth/src/androidTest/java/org/forgerock/android/auth/webauthn/MockWebAuthnAuthenticationCallback.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+/**
+ * Extend the {@link WebAuthnAuthenticationCallback} and provide mocking result
+ * from {@link Fido2ApiClient}
+ */
+public class MockWebAuthnAuthenticationCallback extends WebAuthnAuthenticationCallback {
+
+    public MockWebAuthnAuthenticationCallback() {
+    }
+
+    public MockWebAuthnAuthenticationCallback(JSONObject jsonObject, int index) {
```
Comment (rodrigoareis) [inline forgerock-auth/src/androidTest/java/org/forgerock/android/auth/webauthn/MockWebAuthnAuthenticationCallback.java:32]: Missing javadoc

File: forgerock-auth/src/androidTest/java/org/forgerock/android/auth/webauthn/MockWebAuthnRegistrationCallback.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+/**
+ * Extend the {@link WebAuthnRegistrationCallback} and provide mocking result
+ * from {@link Fido2ApiClient}
+ */
+public class MockWebAuthnRegistrationCallback extends WebAuthnRegistrationCallback {
+
+    public MockWebAuthnRegistrationCallback() {
+    }
+
+    public MockWebAuthnRegistrationCallback(JSONObject jsonObject, int index) {
```
Comment (rodrigoareis) [inline forgerock-auth/src/androidTest/java/org/forgerock/android/auth/webauthn/MockWebAuthnRegistrationCallback.java:32]: Missing javadoc

File: forgerock-auth/src/androidTest/res/xml/forgerock_authenticator.xml [hunk 10]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+  ~ Copyright (c) 2019 ForgeRock. All rights reserved.
```
Comment (rodrigoareis) [inline forgerock-auth/src/androidTest/res/xml/forgerock_authenticator.xml:3]: Update copyright

File: forgerock-auth/src/debug/java/org/forgerock/android/auth/DummyActivity.java [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+ * of the MIT license. See the LICENSE file for details.
+ */
+
+package org.forgerock.android.auth;
+
+import android.os.Bundle;
+
+import androidx.fragment.app.FragmentActivity;
+
+public class DummyActivity extends FragmentActivity {
```
Comment (rodrigoareis) [inline forgerock-auth/src/debug/java/org/forgerock/android/auth/DummyActivity.java:14]: Missing javadoc

File: forgerock-auth/src/main/java/org/forgerock/android/auth/InitProvider.java [hunk 12]
```diff
@@ -32,6 +33,10 @@ public static Activity getCurrentActivity() {
         return currentActivity.get();
     }
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/InitProvider.java:35]: Missing javadoc

File: forgerock-auth/src/main/java/org/forgerock/android/auth/exception/WebAuthnException.java [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+ */
+
+package org.forgerock.android.auth.exception;
+
+/**
+ * Exception that is thrown when the webauthn registration or authentication failed
+ */
+public class WebAuthnException extends Exception {
+
+    public WebAuthnException(String message) {
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/exception/WebAuthnException.java:15]: Missing javadoc

File: forgerock-auth/src/main/java/org/forgerock/android/auth/exception/WebAuthnResponseException.java [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+/**
+ * An Exception representation of {@link AuthenticatorErrorResponse}
+ */
+@Getter
+public class WebAuthnResponseException extends Exception {
+
+    private final ErrorCode errorCode;
+    private final int errorCodeAsInt;
+
+    public WebAuthnResponseException(AuthenticatorErrorResponse authenticatorErrorResponse) {
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/exception/WebAuthnResponseException.java:24]: Missing javadoc

File: forgerock-auth/src/main/java/org/forgerock/android/auth/webauthn/PublicKeyCredentialSource.java [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+import java.util.Arrays;
+import java.util.Collections;
+
+import lombok.Builder;
+import lombok.Getter;
+
+import static com.google.android.gms.fido.common.Transport.INTERNAL;
+
+@Getter
+public class PublicKeyCredentialSource implements Parcelable {
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/webauthn/PublicKeyCredentialSource.java:29]: Missing javadoc

File: forgerock-auth/src/main/java/org/forgerock/android/auth/webauthn/WebAuthKeySelectionDialogFragment.java [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+ * A Simple Dialog to display key selection using a spinner
+ */
+public class WebAuthKeySelectionDialogFragment extends DialogFragment {
+
+    private static final String PUBLIC_KEY_CREDENTIAL_SOURCE = "PUBLIC_KEY_CREDENTIAL_SOURCE";
+    private boolean isCancel = true;
+    private List<PublicKeyCredentialSource> sources;
+    private FRListener<PublicKeyCredentialSource> listener;
+
+    public WebAuthKeySelectionDialogFragment() {
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/webauthn/WebAuthKeySelectionDialogFragment.java:43]: Should we provide a javadoc for this constructor?

File: forgerock-auth/src/main/java/org/forgerock/android/auth/webauthn/WebAuthnAuthentication.java [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+            allowCredentials = value.getJSONArray(_ALLOW_CREDENTIALS);
+        } else if (value.has(ALLOW_CREDENTIALS)) {
+            String allowCredentialString = value.getString(ALLOW_CREDENTIALS)
+                    .replaceAll("(allowCredentials: |new Int8Array\\(|\\).buffer )", "");
+            if (allowCredentialString.trim().length() > 0) {
+                allowCredentials = new JSONArray(allowCredentialString);
+            }
+        }
+        return getCredentials(allowCredentials);
+
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/webauthn/WebAuthnAuthentication.java:82]: Extra line does not mach coding style for the rest of methods

File: forgerock-auth/src/main/java/org/forgerock/android/auth/webauthn/WebAuthnAuthentication.java [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+                    }
+                    Listener.onSuccess(listener, sb.toString());
+                }
+
+                @Override
+                public void onException(Exception e) {
+                    onWebAuthnException(listener, e);
+                }
+
+
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/webauthn/WebAuthnAuthentication.java:211]: Extra line

File: forgerock-auth/src/main/java/org/forgerock/android/auth/webauthn/WebAuthnAuthentication.java [hunk 19]
```diff
@@ -0,0 +1,216 @@
...
+                if (webAuthnKeySelector == null) {
+                    //Use the default Fragment Dialog
+                    webAuthnKeySelector = WebAuthnKeySelector.DEFAULT;
+                }
+                webAuthnKeySelector.select(fragmentManager, publicKeyCredentialSources, selectedListener);
+            }
+        } else {
+            authenticate(context, fragmentManager, listener, allowCredentials, null);
+        }
+
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/webauthn/WebAuthnAuthentication.java:157]: Extra line

File: forgerock-auth/src/main/java/org/forgerock/android/auth/webauthn/WebAuthnHeadlessAuthenticateFragment.java [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+                            REQUEST_FIDO2_SIGNIN,
+                            null, 0, 0, 0, null);
+                } catch (IntentSender.SendIntentException e) {
+                    Listener.onException(listener, new WebAuthnException(e));
+                }
+            }
+        }
+    }
+
+    public void setListener(FRListener<AuthenticatorAssertionResponse> listener) {
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/webauthn/WebAuthnHeadlessAuthenticateFragment.java:76]: Missing javadoc

File: forgerock-auth/src/main/java/org/forgerock/android/auth/webauthn/WebAuthnHeadlessRegistrationFragment.java [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+                            REQUEST_FIDO2_REGISTER,
+                            null, 0, 0, 0, null);
+                } catch (IntentSender.SendIntentException e) {
+                    Listener.onException(listener, new WebAuthnException(e));
+                }
+            }
+        }
+    }
+
+    public void setListener(FRListener<AuthenticatorAttestationResponse> listener) {
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/webauthn/WebAuthnHeadlessRegistrationFragment.java:78]: Missing javadoc

File: forgerock-auth/src/androidTest/java/org/forgerock/android/auth/webauthn/MockWebAuthnAuthenticationCallback.java (hunk 8 above)
Comment (witrisna) [reply forgerock-auth/src/androidTest/java/org/forgerock/android/auth/webauthn/MockWebAuthnAuthenticationCallback.java:32]: I would skip this javadoc for the testing class

File: forgerock-auth/src/androidTest/java/org/forgerock/android/auth/webauthn/MockWebAuthnRegistrationCallback.java (hunk 9 above)
Comment (witrisna) [reply forgerock-auth/src/androidTest/java/org/forgerock/android/auth/webauthn/MockWebAuthnRegistrationCallback.java:32]: I would skip this javadoc for the testing class

File: forgerock-auth/src/test/java/org/forgerock/android/auth/AccountDataRepositoryTest.java [hunk 22]
```diff
@@ -0,0 +1,17 @@
+/*
+ * Copyright (c) 2021 ForgeRock. All rights reserved.
+ *
+ * This software may be modified and distributed under the terms
+ * of the MIT license. See the LICENSE file for details.
+ */
+
+package org.forgerock.android.auth;
+
+public class AccountDataRepositoryTest extends DataRepositoryTest {
```
Comment (rodrigoareis) [inline forgerock-auth/src/test/java/org/forgerock/android/auth/AccountDataRepositoryTest.java:10]: Missing javadoc

File: forgerock-auth/src/test/java/org/forgerock/android/auth/AuthTestSuite.java [hunk 23]
```diff
@@ -8,6 +8,8 @@
 package org.forgerock.android.auth;
 
 import org.forgerock.android.auth.callback.*;
+import org.forgerock.android.auth.webauthn.WebAuthnAuthenticationTest;
```
Comment (rodrigoareis) [inline forgerock-auth/src/test/java/org/forgerock/android/auth/AuthTestSuite.java:11]: Update copyright

File: forgerock-auth/src/test/java/org/forgerock/android/auth/DummyTask.java [hunk 24]
```diff
@@ -0,0 +1,92 @@
...
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import com.google.android.gms.tasks.OnFailureListener;
+import com.google.android.gms.tasks.OnSuccessListener;
+import com.google.android.gms.tasks.Task;
+
+import java.util.concurrent.Executor;
+
+public class DummyTask extends Task<PendingIntent> {
```
Comment (rodrigoareis) [inline forgerock-auth/src/test/java/org/forgerock/android/auth/DummyTask.java:22]: Missing javadoc

File: forgerock-auth/src/test/java/org/forgerock/android/auth/MockWebAuthnRegistrationCallback.java [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+import com.google.android.gms.fido.fido2.api.common.PublicKeyCredentialCreationOptions;
+import com.google.android.gms.tasks.OnSuccessListener;
+import com.google.android.gms.tasks.Task;
+
+import org.forgerock.android.auth.callback.WebAuthnRegistrationCallback;
+import org.forgerock.android.auth.webauthn.WebAuthnRegistration;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+public class MockWebAuthnRegistrationCallback extends WebAuthnRegistrationCallback {
```
Comment (rodrigoareis) [inline forgerock-auth/src/test/java/org/forgerock/android/auth/MockWebAuthnRegistrationCallback.java:24]: Missing javadoc

File: forgerock-auth/src/test/java/org/forgerock/android/auth/SharedPreferencesDataRepositoryTest.java [hunk 26]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * This software may be modified and distributed under the terms
+ * of the MIT license. See the LICENSE file for details.
+ */
+
+package org.forgerock.android.auth;
+
+import android.content.SharedPreferences;
+
+public class SharedPreferencesDataRepositoryTest extends DataRepositoryTest {
```
Comment (rodrigoareis) [inline forgerock-auth/src/test/java/org/forgerock/android/auth/SharedPreferencesDataRepositoryTest.java:12]: Missing javadoc

File: forgerock-auth/src/test/java/org/forgerock/android/auth/WebAuthnFlowTest.java [hunk 27]
```diff
@@ -0,0 +1,143 @@
...
+                    throw new RuntimeException(e);
+                }
+                state.next(context, this);
+            }
+        };
+        frAuth.next(context, nodeListenerFuture);
+        nodeListenerFuture.get();
+    }
+
+    /*
```
Comment (rodrigoareis) [inline forgerock-auth/src/test/java/org/forgerock/android/auth/WebAuthnFlowTest.java:91]: Should those commented tests be removed?

File: forgerock-auth/src/test/java/org/forgerock/android/auth/callback/MetadataCallbackTest.java [hunk 28]
```diff
@@ -7,6 +7,7 @@
 
 package org.forgerock.android.auth.callback;
 
+import org.forgerock.android.auth.BaseTest;
```
Comment (rodrigoareis) [inline forgerock-auth/src/test/java/org/forgerock/android/auth/callback/MetadataCallbackTest.java:10]: Update copyright

File: forgerock-core/src/main/java/org/forgerock/android/auth/Encryptor.java [hunk 29]
```diff
@@ -61,4 +61,20 @@ static Encryptor getEncryptor(Context context, String keyAlias, SecretKeyStore s
         }
     }
 
+    @SuppressLint("NewApi")
```
Comment (rodrigoareis) [inline forgerock-core/src/main/java/org/forgerock/android/auth/Encryptor.java:64]: Update copyright

File: samples/auth/build.gradle [hunk 30]
```diff
@@ -32,6 +31,15 @@ android {
         }
     }
 
+    signingConfigs {
```
Comment (rodrigoareis) [inline samples/auth/build.gradle:34]: Update copyright

File: samples/auth/src/main/java/org/forgerock/auth/MainActivity.java [hunk 31]
```diff
@@ -352,8 +352,9 @@ public void launchBrowser() {
         FRUser.browser().appAuthConfigurer()
                 .authorizationRequest(r -> {
                     Map<String, String> additionalParameters = new HashMap<>();
-                    additionalParameters.put("acr_values", "exampletree");
+                    additionalParameters.put("service", "Simple");
```
Comment (rodrigoareis) [inline samples/auth/src/main/java/org/forgerock/auth/MainActivity.java:355]: Update copyright

File: samples/auth/src/main/res/values/strings.xml [hunk 32]
```diff
@@ -30,7 +29,7 @@
 
     <!-- The following setting is for local development environment only -->
     <!-- Server -->
-    <string name="forgerock_url" translatable="false">http://openam.example.com:8081/openam</string>
+    <string name="forgerock_url" translatable="false">http://openam.example.com/openam</string>
```
Comment (rodrigoareis) [inline samples/auth/src/main/res/values/strings.xml:32]: Update copyright

File: samples/authenticator/build.gradle [hunk 33]
```diff
@@ -32,6 +32,10 @@ android {
         targetCompatibility JavaVersion.VERSION_1_8
     }
 
+    lintOptions {
```
Comment (rodrigoareis) [inline samples/authenticator/build.gradle:35]: Update copyright

**Review by rodrigoareis — CHANGES_REQUESTED** [review]
Overall implementation looks good. Some updates are required:
- Missing some copyright updates
- Few coding style inconsistence
- Some missing javadoc on public methods
- I think for any support test class we should provide javadoc. Those are not required for classes including unit tests 
- Should we commit the debug key (samples/debug.jks) store in the repository?

File: forgerock-auth/src/main/java/org/forgerock/android/auth/webauthn/WebAuthKeySelectionDialogFragment.java (hunk 16 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/webauthn/WebAuthKeySelectionDialogFragment.java:43]: Usually I don't, but no harm

File: forgerock-auth/src/test/java/org/forgerock/android/auth/AccountDataRepositoryTest.java (hunk 22 above)
Comment (witrisna) [reply forgerock-auth/src/test/java/org/forgerock/android/auth/AccountDataRepositoryTest.java:10]: I would skip the javadoc for test.

**Review by rodrigoareis — DISMISSED** [review]
Changes looks good to me

**Review by spetrov — APPROVED** [review]
Looks good to me! 👍 
Approved!

**Pull request merged**